fenicsprecice/fenicsprecice.py

```diff
@@ -273,9 +286,47 @@ def initialize(self, coupling_subdomain, mesh, function_space, write_function=No
             Recommended time step value from preCICE.
         """
 
-        coords = function_space.tabulate_dof_coordinates()
+        write_function_space, write_function = None, None
+        if type(write_object) is Function:
+            write_function_space = write_object.function_space()
+            write_function = write_object
+        elif type(write_object) is FunctionSpace:
+            write_function_space = write_object
+            write_function = None
+        elif write_object is None:
+            print("Participant {} is read-only participant".format(self._config.get_participant_name()))
+        else:
+            raise Exception("Given write object is neither of type dolfin.functions.function.Function or "
+                            "dolfin.functions.functionspace.FunctionSpace")
+
+        if read_function_space is None and write_function_space:
+            self._coupling_type = CouplingMode.UNIDIR_WRITE
+            assert(self._config.get_write_data_name())
+            print("Participant {} is write-only participant".format(self._config.get_participant_name()))
+        elif read_function_space and write_function_space is None:
+            self._coupling_type = CouplingMode.UNIDIR_READ
+            assert(self._config.get_read_data_name())
+        elif read_function_space and write_function_space:
+            self._coupling_type = CouplingMode.BIDIR
+            assert(self._config.get_read_data_name() and self._config.get_write_data_name())
+        else:
+            raise Exception("Incorrect read and write function space combination provided. Please check input "
+                            "parameters in initialization")
+
+        if self._coupling_type is CouplingMode.UNIDIR_READ or self._coupling_type is CouplingMode.BIDIR:
+            self._read_function_type = determine_function_type(read_function_space)
+            self._read_function_space = read_function_space
+
+        if self._coupling_type is CouplingMode.UNIDIR_WRITE or self._coupling_type is CouplingMode.BIDIR:
+            self._write_function_type = determine_function_type(write_function_space)
+
+        coords = self._read_function_space.tabulate_dof_coordinates()
         _, self._fenics_dimensions = coords.shape
 
+        # Ensure that function spaces of read and write functions use the same mesh
+        if self._coupling_type is CouplingMode.BIDIR:
+            assert(self._read_function_space.mesh() is write_function_space.mesh())
```

Comment:
```suggestion
        if self._coupling_type is CouplingMode.BIDIR:
            # Ensure that function spaces of read and write functions use the same mesh
            assert(self._read_function_space.mesh() is write_function_space.mesh()), "read_function_space and write_function_space must use the same mesh."
            mesh = self._read_function_space.mesh()
        elif self._coupling_type is CouplingMode.UNIDIR_READ:
            mesh = self._read_function_space.mesh()
        elif self._coupling_type is CouplingMode.UNIDIR_WRITE:
            mesh = self._write_function_space.mesh()
        else:
            raise Exception("Unexpected control flow.")
```
Please add some meaningful error message after the assertion.